src/optable.ts

```diff
@@ -984,19 +1004,35 @@ optable[opcodes.GET_ITER] = function(f: Py_FrameObject, t: Thread) {
     f.push(builtins.iter(t, f, [tos], new Py_Dict()));
 }
 
-optable[opcodes.FOR_ITER] = function(f: Py_FrameObject) {
-    // calls next() on the iter object at TOS
+optable[opcodes.FOR_ITER] = function(f: Py_FrameObject, t: Thread) {
     var delta = f.readArg();
     var iter = <Iterator> f.peek();
-    var res = iter.next();
-    if (res != null) {
-        f.push(res);
-    } else {
-        f.pop();
-        f.lastInst += delta;
+    // Calls $__next__() if it exists, otherwise, next() on the iterator
+    if (iter.$__next__) {
+        iter.$__next__.exec_from_native(t, f, [], new Py_Dict(), (rv: IPy_Object) => {
+            t.setStatus(ThreadStatus.RUNNABLE);
+        });
+        f.returnToThread = true;
+    }
+    else {
+        var res = iter.next();
+        if (res != null) {
+            f.push(res);
+        } else {
+            f.pop();
+            f.lastInst += delta;
+        }
     }
 }
 
+optable[opcodes.YIELD_VALUE] = function(f: Py_FrameObject, t: Thread) {
+    var res = f.peek();
+    // Push on calling frame's stack
```

Comment:
I think you should just call `asyncReturn` here. `YIELD_VALUE` is essentially identical to `RETURN_VALUE`, except it leaves the generator in a usable state.